HIP-3.md

```diff
@@ -0,0 +1,223 @@
+- It's hard for applications to pay for message fees on their user's behalf. Some contracts are willing to keep a native token balance that they'll use to pay for message fees. However if the fee amount is calculated off-chain and passed into the contract by a user, there's nothing stopping a user from specifying a very high fee and depleting the contract's balance. This approach of contracts paying for fees on a user's behalf is one way of supporting callback messages.
+- Developers are required to use the Hyperlane SDK to estimate correct payments.
+- A transaction won't revert if there is insufficient payment for a message that is being dispatched. With on chain fee quoting, the transaction can be reverted if there is insufficient payment, creating a very strong feedback loop to users. Sending a message should “just work.”
+- Generally, it's a rough user experience. Application developers have an easier time reasoning about and dealing with on-chain data, and want to have certainty that messages will be processed. Requiring off-chain fee calculation limits the composability of Hyperlane.
+
+### **Tech Spec**
+
+Paying for fees can be done by paying an `InterchainGasPaymaster` contract relating to an off-chain relayer that will enter a social contract to process a message. This contract is specific to a relayer operator who has the ability to claim fees paid to the contract.
+
+At a high level, the `InterchainGasPaymaster` contract should accept the message ID of a Hyperlane message that has been dispatched, information about how much gas the relayer should use when processing the message, and payment in the form of native tokens.
+
+Below is the proposed generic interface for an `InterchainGasPaymaster`. Note that implementations of an IGP may place differring semantic meaning to each of the parameters.
+
+```solidity
+interface IInterchainGasPaymaster {
+    /**
+     * @notice Accepts native tokens to pay for the destination gas cost of a
+     *   message. Implementations may or may not require specific payment amounts.
+     * @param _messageId The message ID.
+     * @param _destinationDomain The destination domain of the message.
+     * @param _gas An amount of gas.
+     * @param _refundAddress The address to refund any overpayment to.
+     */
+    function payForGas(
+        bytes32 _messageId,
+        uint32 _destinationDomain,
+        uint256 _gas,
+        address _refundAddress
+    ) external payable;
+}
+```
+
+An IGP that quotes and requires on-chain payment for a total amount of gas that is to be used by the ISM and recipient `handle` would look like:
+
+```
+interface FeeQuotingIgp {
+    /**
+     * @notice Emitted when a gas payment is made.
+     * @param messageId The message ID.
+     * @param amount The amount of native tokens paid.
+     * @param gas The amount of destination gas paid for.
+     */
+    event GasPayment(
```

Comment:
Flagging that this event does not handle unknown chains gracefully

Reply:
Would not charging anything for domains it doesn't know any info about or charging some sort of default fee satisfy this for you? as suggested in https://github.com/hyperlane-xyz/hips/pull/3/files#r1073804672